google/ggrc-core
# GGRC-6563 Display hierarchical assessment role for comments

# Issue description

Assignee roles associated with an assessment have to be displayed based on proposed hierarchy, when a comment is added to an assessment.

# Steps to test the changes

1. Log in to ggrc application.
2. Create an assessment and let the user have one or more roles based on below test cases.
3. Add a comment and verify if the role displayed is according to hierarchy : Verifier, Assignee, Creator, Primary Contact, Secondary Contact, Custom Role*.

Test Cases
1. User is Verifier > Comment > Verifier role
2. User is Assignee > Comment > Assignee role
3. User is Creator > Comment > Creator role
4. User is Creator, Verifier > Comment > Verifier role
5. User is Creator, Assignee > Comment > Assignee role
6. User is Creator, Assignee, Verifier > Comment > Verifier role
7. User is Primary Contact > Comment > Primary Contact role
8. User is Primary Contact, Assignee, Creator > Comment > Assignee role


# Solution description

1. Defined functions 'orderRolesByHierarchy', and 'getAssessmentUserType' in file 'src/ggrc-client/js/plugins/ggrc_utils.js'. These functions order the roles of assessment and change the assignee type as required, when a comment is added from front end.
2. Defined the function '_order_assignee_type_by_role' in file 'src/ggrc/services/resources/assessment.py'. It orders the assignee type when fetched from backend. It is helpful for ordering comments, which were imported.
3. Modified test expected output in file 'src/ggrc-client/js/plugins/tests/ggrc_utils_spec.js' based on proposed assessment roles hierarchy.

# Sanity checklist




# PR Review checklist
